crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -0,0 +1,1022 @@
+        // And is the signature in the device info valid?
+        let maybe_master_key_info = self
+            .master_key_if_device_is_signed_by_user(&sender_device, &sender_user_identity)
+            .await?;
+
+        if let Some((master_key, master_key_verified)) = maybe_master_key_info {
+            // Yes: H (cross-signing key matches that used to sign the device info!)
+            // and: J (device info is verified by matching cross-signing key)
+
+            // Find the actual key within the MasterPubkey struct
+            if let Some(master_key) = master_key.get_first_key() {
+                // We have user_id and master_key for the user sending the to-device message.
+                // Decide the master_key trust level based on whether we have verified this
+                // user. Set the user_id, master_key and trust level in the
+                // session.
+                Ok(SenderData::SenderKnown {
+                    user_id: sender_device.user_id().to_owned(),
+                    master_key,
+                    master_key_verified,
+                })
+            } else {
+                // Surprisingly, there was no key in the MasterPubkey. We did not expect this:
+                // treat it as if the device was not signed by this master key.
+                //
+                tracing::error!(
+                    "MasterPubkey for user {} does not contain any keys!",
+                    master_key.user_id()
+                );
+
+                Ok(SenderData::DeviceInfo {
+                    device_keys: sender_device.as_device_keys().clone(),
+                    retry_details: SenderDataRetryDetails::retry_soon(),
+                    legacy_session: true, // TODO: change to false when retries etc. are done
+                })
+            }
+        } else {
+            // No: Device was not signed by the known identity of the sender.
+            // (Or the signature was invalid. We don't know which unfortunately, so we can't
+            // bail out completely if the signature was invalid, as we'd like
+            // to.)
+            //
+            // Since we've already checked there are >1 signatures, we guess
+            // it was signed by a different identity of this user, so we should retry
+            // later, in case either the device info or the user's identity changes.
+            Ok(SenderData::DeviceInfo {
+                device_keys: sender_device.as_device_keys().clone(),
+                retry_details: SenderDataRetryDetails::retry_soon(),
+                legacy_session: true, // TODO: change to false when retries etc. are done
+            })
+        }
+    }
+
+    /// If `sender_device` is correctly cross-signed by `sender_user_identity`,
+    /// return user_identity's master key and whether it is verified.
+    /// Otherwise, return None.
+    async fn master_key_if_device_is_signed_by_user<'i>(
```

Comment:
This can be simplified using `Device::is_cross_signed_by_owner()`:

https://github.com/matrix-org/matrix-rust-sdk/blob/81d388a55bc8041314bf69a872f93333deb3d3a6/crates/matrix-sdk-crypto/src/identities/device.rs#L281-L286

Another reason why I suggested to use `Device` here.

Reply:
Addressed in https://github.com/matrix-org/matrix-rust-sdk/pull/3590#discussion_r1677761085